haystack_experimental/components/agents/agent.py

```diff
@@ -0,0 +1,264 @@
+
+    The component processes messages and executes tools until a exit_condition condition is met.
+    The exit_condition can be triggered either by a direct text response or by invoking a specific designated tool.
+
+    ### Usage example
+    ```python
+    from haystack.dataclasses import ChatMessage
+    from haystack.tools.tool import Tool
+
+    tools = [Tool(name="calculator", description="..."), Tool(name="search", description="...")]
+    input_types = {"search_depth": int}
+    output_types = {"total_tokens": int}
+
+    agent = Agent(
+        model="anthropic:claude-3",
+        generation_kwargs={"temperature": 0.7},
+        tools=tools,
+        exit_condition="search",
+        input_variables=input_types,
+        output_variables=output_types
+    )
+
+    # Run the agent
+    result = agent.run(
+        messages=[ChatMessage.from_user("Find information about Haystack")],
+        search_depth=2
+    )
+
+    assert "messages" in result  # Contains conversation history
+    assert "total_tokens" in result  # Contains tool execution outputs
+    ```
+    """
+
+    def __init__( # pylint: disable=too-many-positional-arguments
+        self,
+        model: str,
+        tools: Optional[List[Tool]] = None,
+        system_prompt: Optional[str] = None,
+        api_key: Secret = Secret.from_env_var("LLM_API_KEY", strict=False),
+        exit_condition: str = "text",
+        generation_kwargs: Optional[Dict[str, Any]] = None,
+        state_schema: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Initialize the agent component.
+
+        :param model: Model identifier in the format "provider:model_name"
+        :param generation_kwargs: Keyword arguments for the chat model generator
+        :param tools: List of Tool objects available to the agent
+        :param exit_condition: Either "text" if the agent should return when it generates a message without tool calls
+            or the name of a tool that will cause the agent to return once the tool was executed
+        :param input_variables: Dictionary mapping input variable names to their types
+        :param output_variables: Dictionary mapping output variable names to their types
+        :raises ValueError: If model string format is invalid or exit_condition is not valid
+        """
+        if ":" not in model:
+            raise ValueError("Model string must be in format 'provider:model_name'")
+
+        provider, _ = model.split(":")
+        if provider not in _PROVIDER_GENERATOR_MAPPING:
+            raise ValueError(
+                f"Provider must be one of {list(_PROVIDER_GENERATOR_MAPPING.keys())}"
+            )
+
+
+        valid_exits = ["text"] + [tool.name for tool in tools or []]
+        if exit_condition not in valid_exits:
+            raise ValueError(f"Exit condition must be one of {valid_exits}")
+
+        if state_schema is not None:
+            _validate_schema(state_schema)
```

Comment:
I was looking at your note

> If a tool returns data that is incompatible with your declared type, or your handler produces a final value that doesn’t match the type, an error or unexpected behavior may occur. Ensure each attribute’s usage and handler align with its declared type.

and I was wondering if it would be possible here to validate the state_schema against the output type of the Tools? Could be something for later since I'm unsure if there is an easy way for us to grab the output types of each Tool.